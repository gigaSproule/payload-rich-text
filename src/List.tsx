import type { TextData } from "./Text";
import { Text } from "./Text";
import type { Options } from "./types";

export type ListItem = {
  indent: number;
  type: "listitem";
  children: TextData[];
  direction: "ltr" | "rtl" | null;
  format: "left" | "start" | "center" | "right" | "end" | "justify" | "";
  version: number;
  value: number;
};

export type BulletListItem = ListItem;

export type NumberListItem = ListItem;

export type CheckListItem = ListItem & {
  checked: boolean;
};

export type ListData = {
  type: "list";
  direction: "ltr" | "rtl" | null;
  format: string;
  indent: number;
  version: number;
  start: number;
};

export type BulletListData = ListData & {
  listType: "bullet";
  children: BulletListItem[];
  tag: "ul";
};

export type NumberListData = ListData & {
  listType: "number";
  children: NumberListItem[];
  tag: "ol";
};

export type CheckListData = ListData & {
  listType: "check";
  children: CheckListItem[];
  tag: "ul";
};

export type Props = {
  data: BulletListData | NumberListData | CheckListData;
  options?: Options;
};

export const List = ({ data, options }: Props) => {
  const populatedChildren = data.children.filter(
    (child) => child.children.length > 0,
  );
  if (populatedChildren.length === 0) {
    return null;
  }
  if (data.listType === "number") {
    const list = (populatedChildren as NumberListData["children"]).flatMap(
      (listChild) => {
        const children = listChild.children.map((listItem) => (
          <Text key={listItem.text} data={listItem} options={options} />
        ));
        return options?.numberListItem ? (
          options.numberListItem(listChild, children)
        ) : (
          <li key={listChild.value} value={listChild.value}>
            {children}
          </li>
        );
      },
    );
    return options?.numberList ? (
      options.numberList(data, list)
    ) : (
      <ol className="list-number">{list}</ol>
    );
  } else if (data.listType === "check") {
    const list = (populatedChildren as CheckListData["children"]).flatMap(
      (listChild) => {
        const children = listChild.children.map((listItem) => (
          <Text key={listItem.text} data={listItem} options={options} />
        ));
        const id = crypto.randomUUID();
        return options?.checkListItem ? (
          options.checkListItem(listChild, children)
        ) : (
          <li
            key={listChild.value}
            aria-checked={listChild.checked}
            role="checkbox"
            tabIndex={-1}
            value={listChild.value}
          >
            <input id={id} type="checkbox" defaultChecked={listChild.checked} />
            <label htmlFor={id}>{children}</label>
          </li>
        );
      },
    );
    return options?.checkList ? (
      options.checkList(data, list)
    ) : (
      <ul className="list-check">{list}</ul>
    );
  } else {
    const list = (populatedChildren as BulletListData["children"]).flatMap(
      (listChild) => {
        const children = listChild.children.map((listItem) => (
          <Text key={listItem.text} data={listItem} options={options} />
        ));
        return options?.bulletListItem ? (
          options.bulletListItem(listChild, children)
        ) : (
          <li key={listChild.value} value={listChild.value}>
            {children}
          </li>
        );
      },
    );
    return options?.bulletList ? (
      options.bulletList(data, list)
    ) : (
      <ul className="list-bullet">{list}</ul>
    );
  }
};
