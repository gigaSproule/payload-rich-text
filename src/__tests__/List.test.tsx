import { afterAll, beforeAll, describe, expect, it, jest } from "@jest/globals";
import { render } from "@testing-library/react";
import type { Props } from "../List";
import { List } from "../List";
import {
  createBulletListData,
  createBulletListItem,
  createCheckListData,
  createCheckListItem,
  createNumberListData,
  createNumberListItem,
} from "./helpers/createListData";

const originalCrypto = global.crypto;

beforeAll(() => {
  Object.defineProperty(global, "crypto", {
    value: {
      randomUUID: jest.fn().mockReturnValue("randomUUID"),
    },
    writable: true,
  });
});

afterAll(() => {
  Object.defineProperty(global, "crypto", {
    value: originalCrypto,
    writable: true,
  });
});

describe("List", () => {
  it("should not render anything if no children provided", () => {
    const listData: Props["data"] = createBulletListData({ children: [] });
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });
});

describe("NumberList", () => {
  it("should not render anything if no sub-children provided", () => {
    const listData: Props["data"] = createNumberListData({
      children: [createNumberListItem({ children: [] })],
    });
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ol with default li if no options provided", () => {
    const listData: Props["data"] = createNumberListData();
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ol with default li if options provided without numberList and without numberListItem", () => {
    const listData: Props["data"] = createNumberListData();
    const { container } = render(<List data={listData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should render custom list render with default li if options provided with numberList and without numberListItem", () => {
    const listData: Props["data"] = createNumberListData();
    const { container } = render(
      <List
        data={listData}
        options={{
          numberList: (numberList, children) => {
            return (
              <div>
                <span>Custom Number List - {JSON.stringify(numberList)}</span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render custom list item render with default ol if options provided without numberList and with numberListItem", () => {
    const listData: Props["data"] = createNumberListData();
    const { container } = render(
      <List
        data={listData}
        options={{
          numberListItem: (numberListItem, children) => {
            return (
              <div key={numberListItem.value}>
                <span>
                  Custom Number List Item - {JSON.stringify(numberListItem)}
                </span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe("CheckList", () => {
  it("should not render anything if no sub-children provided", () => {
    const listData: Props["data"] = createCheckListData({
      children: [createCheckListItem({ children: [] })],
    });
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ul with default li if no options provided", () => {
    const listData: Props["data"] = createCheckListData();
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ul with default li if no options provided and list item checked", () => {
    const listData: Props["data"] = createCheckListData({
      children: [createCheckListItem({ checked: true })],
    });
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ul with default li if options provided without checkList and without checkListItem", () => {
    const listData: Props["data"] = createCheckListData();
    const { container } = render(<List data={listData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should render custom list render with default li if options provided with checkList and without checkListItem", () => {
    const listData: Props["data"] = createCheckListData();
    const { container } = render(
      <List
        data={listData}
        options={{
          checkList: (checkList, children) => {
            return (
              <div>
                <span>Custom Check List - {JSON.stringify(checkList)}</span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render custom list item render with default ul if options provided without checkList and with checkListItem", () => {
    const listData: Props["data"] = createCheckListData();
    const { container } = render(
      <List
        data={listData}
        options={{
          checkListItem: (checkListItem, children) => {
            return (
              <div key={checkListItem.value}>
                <span>
                  Custom Check List Item - {JSON.stringify(checkListItem)}
                </span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe("BulletList", () => {
  it("should not render anything if no sub-children provided", () => {
    const listData: Props["data"] = createBulletListData({
      children: [createBulletListItem({ children: [] })],
    });
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ul with default li if no options provided", () => {
    const listData: Props["data"] = createBulletListData();
    const { container } = render(<List data={listData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default ul with default li if options provided without bulletList and without bulletListItem", () => {
    const listData: Props["data"] = createBulletListData();
    const { container } = render(<List data={listData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should render custom list render with default li if options provided with bulletList and without bulletListItem", () => {
    const listData: Props["data"] = createBulletListData();
    const { container } = render(
      <List
        data={listData}
        options={{
          bulletList: (bulletList, children) => {
            return (
              <div>
                <span>Custom Bullet List - {JSON.stringify(bulletList)}</span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render custom list item render with default ul if options provided without bulletList and with bulletListItem", () => {
    const listData: Props["data"] = createBulletListData();
    const { container } = render(
      <List
        data={listData}
        options={{
          bulletListItem: (bulletListItem, children) => {
            return (
              <div key={bulletListItem.value}>
                <span>
                  Custom Bullet List Item - {JSON.stringify(bulletListItem)}
                </span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
