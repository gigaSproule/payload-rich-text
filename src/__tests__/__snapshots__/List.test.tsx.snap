// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulletList should not render anything if no sub-children provided 1`] = `<div />`;

exports[`BulletList should render custom list item render with default ul if options provided without bulletList and with bulletListItem 1`] = `
<div>
  <ul
    class="list-bullet"
  >
    <div>
      <span>
        Custom Bullet List Item - 
        {"type":"listitem","indent":0,"direction":null,"format":"","version":1,"value":1,"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}]}
      </span>
      heading text
    </div>
  </ul>
</div>
`;

exports[`BulletList should render custom list render with default li if options provided with bulletList and without bulletListItem 1`] = `
<div>
  <div>
    <span>
      Custom Bullet List - 
      {"type":"list","direction":null,"format":"","indent":0,"version":1,"start":1,"listType":"bullet","children":[{"type":"listitem","indent":0,"direction":null,"format":"","version":1,"value":1,"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}]}],"tag":"ul"}
    </span>
    <li
      value="1"
    >
      heading text
    </li>
  </div>
</div>
`;

exports[`BulletList should render default ul with default li if no options provided 1`] = `
<div>
  <ul
    class="list-bullet"
  >
    <li
      value="1"
    >
      heading text
    </li>
  </ul>
</div>
`;

exports[`BulletList should render default ul with default li if options provided without bulletList and without bulletListItem 1`] = `
<div>
  <ul
    class="list-bullet"
  >
    <li
      value="1"
    >
      heading text
    </li>
  </ul>
</div>
`;

exports[`BulletList should render list item link if list item is a link 1`] = `
<div>
  <ul
    class="list-bullet"
  >
    <li
      value="1"
    >
      <a
        href="https://www.example.com"
      >
        heading text
      </a>
    </li>
  </ul>
</div>
`;

exports[`BulletList should render list with sub lists if sub lists are provided 1`] = `
<div>
  <ul
    class="list-bullet"
  >
    <li
      value="1"
    >
      <ul
        class="list-bullet"
      >
        <li
          value="1"
        >
          <ul
            class="list-bullet"
          >
            <li
              value="1"
            >
              <ul
                class="list-bullet"
              >
                <li
                  value="1"
                >
                  heading text
                </li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`CheckList should not render anything if no sub-children provided 1`] = `<div />`;

exports[`CheckList should render custom list item render with default ul if options provided without checkList and with checkListItem 1`] = `
<div>
  <ul
    class="list-check"
  >
    <div>
      <span>
        Custom Check List Item - 
        {"type":"listitem","indent":0,"direction":null,"format":"","version":1,"value":1,"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}],"checked":false}
      </span>
      heading text
    </div>
  </ul>
</div>
`;

exports[`CheckList should render custom list render with default li if options provided with checkList and without checkListItem 1`] = `
<div>
  <div>
    <span>
      Custom Check List - 
      {"type":"list","direction":null,"format":"","indent":0,"version":1,"start":1,"listType":"check","children":[{"type":"listitem","indent":0,"direction":null,"format":"","version":1,"value":1,"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}],"checked":false}],"tag":"ul"}
    </span>
    <li
      aria-checked="false"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        heading text
      </label>
    </li>
  </div>
</div>
`;

exports[`CheckList should render default ul with default li if no options provided 1`] = `
<div>
  <ul
    class="list-check"
  >
    <li
      aria-checked="false"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        heading text
      </label>
    </li>
  </ul>
</div>
`;

exports[`CheckList should render default ul with default li if no options provided and list item checked 1`] = `
<div>
  <ul
    class="list-check"
  >
    <li
      aria-checked="true"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        checked=""
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        heading text
      </label>
    </li>
  </ul>
</div>
`;

exports[`CheckList should render default ul with default li if options provided without checkList and without checkListItem 1`] = `
<div>
  <ul
    class="list-check"
  >
    <li
      aria-checked="false"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        heading text
      </label>
    </li>
  </ul>
</div>
`;

exports[`CheckList should render list item link if list item is a link 1`] = `
<div>
  <ul
    class="list-check"
  >
    <li
      aria-checked="false"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        <a
          href="https://www.example.com"
        >
          heading text
        </a>
      </label>
    </li>
  </ul>
</div>
`;

exports[`CheckList should render list with sub lists if sub lists are provided 1`] = `
<div>
  <ul
    class="list-check"
  >
    <li
      aria-checked="false"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        <ul
          class="list-check"
        >
          <li
            aria-checked="false"
            role="checkbox"
            tabindex="-1"
            value="1"
          >
            <input
              id="randomUUID"
              type="checkbox"
            />
            <label
              for="randomUUID"
            >
              <ul
                class="list-check"
              >
                <li
                  aria-checked="false"
                  role="checkbox"
                  tabindex="-1"
                  value="1"
                >
                  <input
                    id="randomUUID"
                    type="checkbox"
                  />
                  <label
                    for="randomUUID"
                  >
                    <ul
                      class="list-check"
                    >
                      <li
                        aria-checked="false"
                        role="checkbox"
                        tabindex="-1"
                        value="1"
                      >
                        <input
                          id="randomUUID"
                          type="checkbox"
                        />
                        <label
                          for="randomUUID"
                        >
                          heading text
                        </label>
                      </li>
                    </ul>
                  </label>
                </li>
              </ul>
            </label>
          </li>
        </ul>
      </label>
    </li>
  </ul>
</div>
`;

exports[`List should not render anything if no children provided 1`] = `<div />`;

exports[`NumberList should not render anything if no sub-children provided 1`] = `<div />`;

exports[`NumberList should render custom list item render with default ol if options provided without numberList and with numberListItem 1`] = `
<div>
  <ol
    class="list-number"
    style="list-style: decimal;"
  >
    <div>
      <span>
        Custom Number List Item - 
        {"type":"listitem","indent":0,"direction":null,"format":"","version":1,"value":1,"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}]}
      </span>
      heading text
    </div>
  </ol>
</div>
`;

exports[`NumberList should render custom list render with default li if options provided with numberList and without numberListItem 1`] = `
<div>
  <div>
    <span>
      Custom Number List - 
      {"type":"list","direction":null,"format":"","indent":0,"version":1,"start":1,"listType":"number","children":[{"type":"listitem","indent":0,"direction":null,"format":"","version":1,"value":1,"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}]}],"tag":"ol"}
    </span>
    <li
      value="1"
    >
      heading text
    </li>
  </div>
</div>
`;

exports[`NumberList should render default ol with default li if no options provided 1`] = `
<div>
  <ol
    class="list-number"
    style="list-style: decimal;"
  >
    <li
      value="1"
    >
      heading text
    </li>
  </ol>
</div>
`;

exports[`NumberList should render default ol with default li if options provided without numberList and without numberListItem 1`] = `
<div>
  <ol
    class="list-number"
    style="list-style: decimal;"
  >
    <li
      value="1"
    >
      heading text
    </li>
  </ol>
</div>
`;

exports[`NumberList should render list item link if list item is a link 1`] = `
<div>
  <ol
    class="list-number"
    style="list-style: decimal;"
  >
    <li
      value="1"
    >
      <a
        href="https://www.example.com"
      >
        heading text
      </a>
    </li>
  </ol>
</div>
`;

exports[`NumberList should render list with sub lists with different styling, resetting after 5 levels deep by default, if sub lists are provided 1`] = `
<div>
  <ol
    class="list-number"
    style="list-style: decimal;"
  >
    <li
      value="1"
    >
      <ol
        class="list-number"
        style="list-style: upper-alpha;"
      >
        <li
          value="1"
        >
          <ol
            class="list-number"
            style="list-style: lower-alpha;"
          >
            <li
              value="1"
            >
              <ol
                class="list-number"
                style="list-style: upper-roman;"
              >
                <li
                  value="1"
                >
                  <ol
                    class="list-number"
                    style="list-style: lower-roman;"
                  >
                    <li
                      value="1"
                    >
                      <ol
                        class="list-number"
                        style="list-style: decimal;"
                      >
                        <li
                          value="1"
                        >
                          heading text
                        </li>
                      </ol>
                    </li>
                  </ol>
                </li>
              </ol>
            </li>
          </ol>
        </li>
      </ol>
    </li>
  </ol>
</div>
`;
